### PR TITLE
Add type annotations

### PR DESCRIPTION
**Pull request by rth**
This adds Python 3.5+ type annotations.

This helps with auto-completion in editors and for static code analysis with [mypy](https://github.com/python/mypy),
```
mypy --ignore-missing-imports phicore
```

 Here it allowed to detect one (non critical) bug -- see below.

### CODE REVIEW

File: phicore/io.py [hunk 1]
```diff
@@ -31,7 +33,7 @@ def __init__(self, fullpath, mode="r", force=False):
 
         if mode not in ('r', 'r+', 'w', 'w+', 'a', 'a+'):
             raise ValueError('Access type {} unkown. See `open` documentation.'
-                             .format(self.mode))
+                             .format(mode))
```
Comment (rth) [inline phicore/io.py:36]: This happened to be incorrect (`self.mode` was not defined there) and mypy caught that.

**Comment by rth** [comment]
Planning to merge this unless @ajeandet you are against it.

**Comment by ajeandet** [comment]
Hi @rth,
No problem for me! Type annotations are welcome!

**Comment by rth** [comment]
Thanks @ajeandet !

**Pull request merged**